articles/150_roslaunch.md

```diff
@@ -428,62 +428,168 @@ The remapping design document[^static_remapping] goes into detail on how you can
+###### Exit Code
+If the container proces is asked to shutdown due to normal [Termination], then the exit code must be 0.
+If it exits due to an error then exit code must be any other number.
 
-It would be more efficient because you don't need to have a node in each proxy to actually make the Service call to the load that node the user wants to run, and you don't need to maintain a bond between the container and the proxy processes.
+##### Parallel vs Sequential Loading of Node
 
-So I'm interested in what others think about coming up with a standardized executable which can load and configure many nodes at once from command line arguments or a config file.
-</div>
+If it is possible to load multiple nodes in parallel, then it needs to be decided how to load the nodes.
+The container process should load nodes as soon as it is asked.
+It should be up to the launch system to decide whether to load nodes in parallel or sequentially.
+If multiple nodes of the same type are to be launched, then the launch system should load the nodes sequentially so each is able to remap it's name before the next is loaded.
+If they are of different types then the launch system may choose to try to load them in parallel, where the exact order they get loaded is determined by chance or the container process.
 
-#### By Proxy
+##### Registration of Composable Nodes
 
-The other way to implement a process which dynamically instantiates ROS nodes, is to send the request to instantiate a node asynchronously from another process.
-So the "container" process would start up and provide a service (ROS Service or otherwise) that lets external processes request a node be instantiated with a given set of configurations.
+How Composable nodes are registered is not defined by this document.
+Instead, the container process is responsible for knowing how to find nodes it is asked to load.
+For example, a container process might use pluginlib for rclcpp nodes, or python entry points for rclpy nodes.
 
-One form this can take is by providing a "proxy" process which doesn't actually run the node, but instead loads it into an already running container and just stays running until the node is shutdown in the remote container.
-This pattern is used by the "nodelets" in ROS 1.
-It's a useful pattern when you're running nodes by hand but you want them to share a process and don't want to write your own program to do that.
+##### Duplicate Node Handling
 
-There are a few other forms this can take, but the common thread between them is that a process instantiates nodes dynamically based on asynchronous input from external actors (proxy's), and that the configuration for those nodes is communicated through something other than command line arguments and environment variables.
+If a container process is asked to load a node with a full node name matching an existing node, then it must reject the request.
```

Comment:
All comments above about using unique ids rather than full names are of course void if we want to forbid duplicate names. What is the main reason to not allow them? The extra effort of using unique ids seems rather small.

Reply:
I'd say (as I did above before seeing this) that we already assume unique node names in other parts of the system, and depending on how we generate the unique id's (are they unique to the container or within the whole system?) they may or may not be simple to generate, so we should avoid duplicate names. What is the main reason to allow them?

Reply:
> I'll add the unique ID. Even if fully qualified node names are unique, if dynamic remapping allows changing a node's name in the future then a unique id may make it easier to track and unload nodes that change their name.

👍 for the rationale of having the unique id for future dynamic node name changing (but again needs to be clear if it's unique in the container or within the system), and also do you still intend to prevent duplicate node names?

Personally I don't see a reason to allow duplicate names.